proposals/Unified FileSystem API.md

```diff
@@ -0,0 +1,116 @@
+# Unified FileSystem API
+
+Authors:
+- @gozala
+
+Initial PR: TBD <!-- Reference the PR first proposing this document. Oooh, self-reference! -->
+
+<!--
+This template is intended to be used by those who would like to pitch a new project for one of the Web3 Dev project teams to take on. It should contain sufficient detail that others can understand how this project contributes to our team’s mission of  product-market fit for our unified stack of protocols, what is included in scope of the project, where to get started if a project team were to take this on, and any other information relevant for prioritizing this project against others.
+Good project scope aims for ~3-5 engineers for 1-3 months (though feel free to suggest larger-scoped projects anyway). Projects do not include regular day-to-day maintenance and improvement work, e.g. on testing, tooling, validation, code clarity, refactors for future capability, etc.
+-->
+
+### Purpose &amp; impact 
+##### Background &amp; intent
+_Describe the desired state of the world after this project? Why does that matter?_
+<!--
+Outline the status quo, including any relevant context on the problem you’re seeing that this project should solve. Wherever possible, include pains or problems that you’ve seen users experience to help motivate why solving this problem works towards top-line objectives. 
+-->
+
+Today IPFS comes with two sets of FileSystem APIs, one offering global network level view (and is less file systemy) and second local node level (mutable) view. Not only thas is confusing and suboptimal, but neither provide an adequate solution for various use cases:
+
+###### File Tree in DAGs
+
+Sometimes you want to stick tree of files into a DAG. This is reasonable straight forward with `ipfs.add` API if tree structure can be represented as stream of file paths and contents. 
+
+However there are cases where representing a file tree as a steam is impractical (e.g. it happens in response to on various events). Such cases are much better serverd with `ipfs.files` API, however it is also not ideal because tree built ends up in the node local FS path and concurrent tasks could introduce race condititions.
+
+###### Multitenant Nodes
+
+When multiple apllications share an IPFS node they end up with shared local filesystem view, which is impractical because they can introduce race condition and access each others secret files.
+
+
+###### Content Froking
+
+Neither of the existing file system APIs are adequate for forking file tree and performing edits on it. Global filesystem API does not really has write API, and with local filesystem API file tree needs to be assigns a path, which not only requires adding file tree and removing afterwards but is also prone to race conditions.
+
+---
+
+What we need instead is an API similar to `ipfs.files` that can be instantiated with an arbitrary `CID`. This would address all the above use cases.
```

Comment:
Hm.. we can do `ipfs files cp /ipfs/{cid} /sub-mfs` for unixfs DAGs and then mutate ` /sub-mfs`.
Perhaps what is missing is ability to copy non-MFS dags to  MFS, and porcelain for mutating them?

Reply:
I don't think copying into mfs is a good way to go about it. All the use cases I had was about building up a tree and then either adding to network or MFS (or both) in one operation. Adding things to mfs introduces room for race conditions and an overhead which is unnecessary, and it also requires cleaning up after.